server/src/main/java/org/elasticsearch/action/support/broadcast/unpromotable/TransportBroadcastUnpromotableAction.java

```diff
@@ -0,0 +1,112 @@
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+package org.elasticsearch.action.support.broadcast.unpromotable;
+
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.elasticsearch.action.ActionListener;
+import org.elasticsearch.action.ActionListenerResponseHandler;
+import org.elasticsearch.action.ActionResponse;
+import org.elasticsearch.action.support.ActionFilters;
+import org.elasticsearch.action.support.HandledTransportAction;
+import org.elasticsearch.action.support.RefCountingListener;
+import org.elasticsearch.cluster.ClusterState;
+import org.elasticsearch.cluster.node.DiscoveryNode;
+import org.elasticsearch.cluster.service.ClusterService;
+import org.elasticsearch.common.Strings;
+import org.elasticsearch.common.io.stream.Writeable;
+import org.elasticsearch.tasks.Task;
+import org.elasticsearch.transport.TransportChannel;
+import org.elasticsearch.transport.TransportRequestHandler;
+import org.elasticsearch.transport.TransportRequestOptions;
+import org.elasticsearch.transport.TransportResponse;
+import org.elasticsearch.transport.TransportService;
+
+public abstract class TransportBroadcastUnpromotableAction<Request extends BroadcastUnpromotableRequest> extends HandledTransportAction<
+    Request,
+    ActionResponse.Empty> {
+
+    protected Logger logger = LogManager.getLogger(getClass());
+
+    protected final ClusterService clusterService;
+    protected final TransportService transportService;
+
+    protected final String transportUnpromotableAction;
+    protected final String executor;
+
+    protected TransportBroadcastUnpromotableAction(
+        String actionName,
+        ClusterService clusterService,
+        TransportService transportService,
+        ActionFilters actionFilters,
+        Writeable.Reader<Request> requestReader,
+        String executor
+    ) {
+        super(actionName, transportService, actionFilters, requestReader);
+        this.clusterService = clusterService;
+        this.transportService = transportService;
+        this.transportUnpromotableAction = actionName + "[u]";
+        this.executor = executor;
+
+        transportService.registerRequestHandler(transportUnpromotableAction, executor, requestReader, new UnpromotableTransportHandler());
+    }
+
+    protected abstract void unpromotableShardOperation(Task task, Request request, ActionListener<ActionResponse.Empty> listener);
+
+    @Override
+    protected void doExecute(Task task, Request request, ActionListener<ActionResponse.Empty> listener) {
+        try (var listeners = new RefCountingListener(listener.map(v -> ActionResponse.Empty.INSTANCE))) {
+            ActionListener.completeWith(listeners.acquire(), () -> {
+                final ClusterState clusterState = clusterService.state();
```

Comment:
I think we should carry the shard routing table in the request rather than getting it again from here.

Reply:
Hi @DaveCTurner , I have not got this point yet, why? In the cases we have so far, both the refresh and new commit notification call this new action locally, so I do not see why the shard routing table would be different from the cluster state's one. Plus, while the Refresh one calls it internally from within another action, the new commit notification does not need to do that anymore (we save one round-trip to the local node of the primary) and can call this action directly.

Reply:
`clusterService#state()` is mutable, it returns the last-applied state, so it may have changed since the last time we read it during the coordination of the refresh action. Mostly this will be fine, but sometimes it's going to yield a totally different routing table (e.g. the index may even be deleted at this point). I'd prefer the caller to have control over which routing table to use for the routing of this request. The new-commit notification can just get it from the cluster service for sure.

Reply:
Hi @DaveCTurner I see what you mean. But in `TransportShardRefreshAction.UnpromotableReplicasRefreshProxy#onPrimaryOperationComplete()` we'd need to get hold of a shard routing table to pass in the Request. Wouldn't we still get that from the cluster service? Or does Refresh somehow holds the primary's shard routing table?

Reply:
it comes from the `ReplicationGroup`:

```diff
diff --git a/server/src/main/java/org/elasticsearch/action/admin/indices/refresh/TransportShardRefreshAction.java b/server/src/main/java/org/elasticsearch/action/admin/indices/refresh/TransportShardRefreshAction.java
index b5d805f270c..9ab8716ce4a 100644
--- a/server/src/main/java/org/elasticsearch/action/admin/indices/refresh/TransportShardRefreshAction.java
+++ b/server/src/main/java/org/elasticsearch/action/admin/indices/refresh/TransportShardRefreshAction.java
@@ -17,6 +17,7 @@ import org.elasticsearch.action.support.replication.ReplicationResponse;
 import org.elasticsearch.action.support.replication.TransportReplicationAction;
 import org.elasticsearch.client.internal.node.NodeClient;
 import org.elasticsearch.cluster.action.shard.ShardStateAction;
+import org.elasticsearch.cluster.routing.IndexShardRoutingTable;
 import org.elasticsearch.cluster.service.ClusterService;
 import org.elasticsearch.common.inject.Inject;
 import org.elasticsearch.common.io.stream.StreamInput;
@@ -108,11 +109,16 @@ public class TransportShardRefreshAction extends TransportReplicationAction<
     protected class UnpromotableReplicasRefreshProxy extends ReplicasProxy {
 
         @Override
-        public void onPrimaryOperationComplete(BasicReplicationRequest replicaRequest, ActionListener<Void> listener) {
+        public void onPrimaryOperationComplete(
+            BasicReplicationRequest replicaRequest,
+            IndexShardRoutingTable indexShardRoutingTable,
+            ActionListener<Void> listener
+        ) {
             assert primaryRefreshResult.refreshed() : "primary has not refreshed";
             ShardId primary = replicaRequest.shardId();
             UnpromotableShardRefreshRequest unpromotableReplicaRequest = new UnpromotableShardRefreshRequest(
                 primary,
+                indexShardRoutingTable,
                 primaryRefreshResult.generation()
             );
             client.executeLocally(
diff --git a/server/src/main/java/org/elasticsearch/action/admin/indices/refresh/UnpromotableShardRefreshRequest.java b/server/src/main/java/org/elasticsearch/action/admin/indices/refresh/UnpromotableShardRefreshRequest.java
index bd302a3dc6e..31416d77178 100644
--- a/server/src/main/java/org/elasticsearch/action/admin/indices/refresh/UnpromotableShardRefreshRequest.java
+++ b/server/src/main/java/org/elasticsearch/action/admin/indices/refresh/UnpromotableShardRefreshRequest.java
@@ -10,6 +10,7 @@ package org.elasticsearch.action.admin.indices.refresh;
 
 import org.elasticsearch.action.ActionRequestValidationException;
 import org.elasticsearch.action.support.broadcast.unpromotable.BroadcastUnpromotableRequest;
+import org.elasticsearch.cluster.routing.IndexShardRoutingTable;
 import org.elasticsearch.common.io.stream.StreamInput;
 import org.elasticsearch.common.io.stream.StreamOutput;
 import org.elasticsearch.index.engine.Engine;
@@ -23,8 +24,12 @@ public class UnpromotableShardRefreshRequest extends BroadcastUnpromotableReques
 
     private final long segmentGeneration;
 
-    public UnpromotableShardRefreshRequest(final ShardId primaryShardId, long segmentGeneration) {
-        super(primaryShardId);
+    public UnpromotableShardRefreshRequest(
+        final ShardId primaryShardId,
+        IndexShardRoutingTable indexShardRoutingTable,
+        long segmentGeneration
+    ) {
+        super(primaryShardId, indexShardRoutingTable);
         this.segmentGeneration = segmentGeneration;
     }
 
diff --git a/server/src/main/java/org/elasticsearch/action/support/broadcast/unpromotable/BroadcastUnpromotableRequest.java b/server/src/main/java/org/elasticsearch/action/support/broadcast/unpromotable/BroadcastUnpromotableRequest.java
index ceef84c2026..8117226529e 100644
--- a/server/src/main/java/org/elasticsearch/action/support/broadcast/unpromotable/BroadcastUnpromotableRequest.java
+++ b/server/src/main/java/org/elasticsearch/action/support/broadcast/unpromotable/BroadcastUnpromotableRequest.java
@@ -10,6 +10,7 @@ package org.elasticsearch.action.support.broadcast.unpromotable;
 
 import org.elasticsearch.action.ActionRequest;
 import org.elasticsearch.action.ActionRequestValidationException;
+import org.elasticsearch.cluster.routing.IndexShardRoutingTable;
 import org.elasticsearch.common.io.stream.StreamInput;
 import org.elasticsearch.common.io.stream.StreamOutput;
 import org.elasticsearch.index.shard.ShardId;
@@ -24,14 +25,16 @@ import static org.elasticsearch.action.ValidateActions.addValidationError;
 public class BroadcastUnpromotableRequest extends ActionRequest {
 
     protected final ShardId primaryShardId;
+    protected final IndexShardRoutingTable indexShardRoutingTable;
 
     public BroadcastUnpromotableRequest(StreamInput in) throws IOException {
-        super(in);
-        primaryShardId = new ShardId(in);
+        assert false;
+        throw new UnsupportedOperationException("not supported");
     }
 
-    public BroadcastUnpromotableRequest(ShardId primaryShardId) {
+    public BroadcastUnpromotableRequest(ShardId primaryShardId, IndexShardRoutingTable indexShardRoutingTable) {
         this.primaryShardId = primaryShardId;
+        this.indexShardRoutingTable = indexShardRoutingTable;
     }
 
     public ShardId primaryShardId() {
diff --git a/server/src/main/java/org/elasticsearch/action/support/replication/ReplicationOperation.java b/server/src/main/java/org/elasticsearch/action/support/replication/ReplicationOperation.java
index 98e1f02c324..f814cb915e1 100644
--- a/server/src/main/java/org/elasticsearch/action/support/replication/ReplicationOperation.java
+++ b/server/src/main/java/org/elasticsearch/action/support/replication/ReplicationOperation.java
@@ -133,9 +133,12 @@ public class ReplicationOperation<
                 logger.trace("[{}] op [{}] completed on primary for request [{}]", primary.routingEntry().shardId(), opType, request);
             }
 
+            final ReplicationGroup replicationGroup = primary.getReplicationGroup();
+
             pendingActions.incrementAndGet();
             replicasProxy.onPrimaryOperationComplete(
                 replicaRequest,
+                replicationGroup.getRoutingTable(),
                 ActionListener.wrap(ignored -> decPendingAndFinishIfNeeded(), e -> { finishAsFailed(e); })
             );
 
@@ -152,7 +155,6 @@ public class ReplicationOperation<
             // on.
             final long maxSeqNoOfUpdatesOrDeletes = primary.maxSeqNoOfUpdatesOrDeletes();
             assert maxSeqNoOfUpdatesOrDeletes != SequenceNumbers.UNASSIGNED_SEQ_NO : "seqno_of_updates still uninitialized";
-            final ReplicationGroup replicationGroup = primary.getReplicationGroup();
             final PendingReplicationActions pendingReplicationActions = primary.getPendingReplicationActions();
             markUnavailableShardsAsStale(replicaRequest, replicationGroup);
             performOnReplicas(replicaRequest, globalCheckpoint, maxSeqNoOfUpdatesOrDeletes, replicationGroup, pendingReplicationActions);
@@ -615,7 +617,11 @@ public class ReplicationOperation<
          * @param replicaRequest             the operation that will be performed on replicas
          * @param listener                   callback for handling the response or failure
          */
-        default void onPrimaryOperationComplete(RequestT replicaRequest, ActionListener<Void> listener) {
+        default void onPrimaryOperationComplete(
+            RequestT replicaRequest,
+            IndexShardRoutingTable indexShardRoutingTable,
+            ActionListener<Void> listener
+        ) {
             listener.onResponse(null);
         }
     }
diff --git a/server/src/test/java/org/elasticsearch/action/support/broadcast/unpromotable/TransportBroadcastUnpromotableActionTests.java b/server/src/test/java/org/elasticsearch/action/support/broadcast/unpromotable/TransportBroadcastUnpromotableActionTests.java
index 553b212d83d..6391ac591f5 100644
--- a/server/src/test/java/org/elasticsearch/action/support/broadcast/unpromotable/TransportBroadcastUnpromotableActionTests.java
+++ b/server/src/test/java/org/elasticsearch/action/support/broadcast/unpromotable/TransportBroadcastUnpromotableActionTests.java
@@ -14,6 +14,7 @@ import org.elasticsearch.action.support.ActionFilters;
 import org.elasticsearch.action.support.ActionTestUtils;
 import org.elasticsearch.action.support.PlainActionFuture;
 import org.elasticsearch.cluster.ClusterState;
+import org.elasticsearch.cluster.routing.IndexShardRoutingTable;
 import org.elasticsearch.cluster.routing.ShardRouting;
 import org.elasticsearch.cluster.routing.ShardRoutingState;
 import org.elasticsearch.cluster.service.ClusterService;
@@ -127,8 +128,8 @@ public class TransportBroadcastUnpromotableActionTests extends ESTestCase {
             super(in);
         }
 
-        TestBroadcastUnpromotableRequest(ShardId primaryShardId) {
-            super(primaryShardId);
+        TestBroadcastUnpromotableRequest(ShardId primaryShardId, IndexShardRoutingTable indexShardRoutingTable) {
+            super(primaryShardId, indexShardRoutingTable);
         }
 
     }
@@ -178,7 +179,10 @@ public class TransportBroadcastUnpromotableActionTests extends ESTestCase {
             ActionTestUtils.execute(
                 broadcastUnpromotableAction,
                 null,
-                new TestBroadcastUnpromotableRequest(primaryShardId.shardId()),
+                new TestBroadcastUnpromotableRequest(
+                    primaryShardId.shardId(),
+                    state.routingTable().shardRoutingTable(primaryShardId.shardId())
+                ),
                 response
             );
         });
@@ -263,7 +267,12 @@ public class TransportBroadcastUnpromotableActionTests extends ESTestCase {
         ShardId wrongShardId = new ShardId(shardId.getIndex(), shardId.getId() + randomIntBetween(10, 100));
         PlainActionFuture<ActionResponse.Empty> response = PlainActionFuture.newFuture();
         logger.debug("--> executing for primary shard id: {}", wrongShardId);
-        ActionTestUtils.execute(broadcastUnpromotableAction, null, new TestBroadcastUnpromotableRequest(wrongShardId), response);
+        ActionTestUtils.execute(
+            broadcastUnpromotableAction,
+            null,
+            new TestBroadcastUnpromotableRequest(wrongShardId, state.routingTable().shardRoutingTable(wrongShardId)),
+            response
+        );
 
         ExecutionException e = expectThrows(ExecutionException.class, () -> response.get());
         assertThat(e, hasToString(containsString("no such shard")));

```

Reply:
We'd still be missing the nodes here - I think it would be reasonable to capture them in the `ReplicationGroup` too, but I would also be ok with grabbing them from the cluster service, with suitable protections for the case where a node doesn't exist.

Reply:
Thanks for the explanation @DaveCTurner . I did not realize ReplicationGroup had the routing table. I guess we'd need to also make IndexShardRoutingTable Writeable. I saw above you did not make it, but I think it needs to be, right? Or do you mean we don't need to transmit it, and it's only active on the local request and then not passed through the wire requests (making it null)?

Reply:
I think we can assert that we never serialise this, it's only ever used locally. It is already kinda writeable via `org.elasticsearch.cluster.routing.IndexShardRoutingTable.Builder#writeTo` but this isn't important. I'm talking with the core/infra folks about dropping the need for these local-only requests to implement `TransportRequest`.

Reply:
Hi @DaveCTurner , it's only used locally but the Request is broadcasted to the unpromotable shards. For example in the case of UnpromotableShardRefreshRequest it has the segmentGeneration. So the whole Request (inheriting from BroadcastUnpromotableRequest) needs to be able to be  serialized. We could drop though the IndexShardRoutingTable from the serialization, thus it becomes null across the wire. Is that fine?

Reply:
Oh ok I see. Hmm. We should really have a different request class for the ones that do go over the wire. I can live with the routing table just being dropped tho.